src/http_server.cc

```diff
@@ -80,6 +80,17 @@ namespace triton { namespace server {
     return;                                         \
   } while (false)
 
+#define RETURN_AND_RESPOND_IF_RESTRICTED(REQ, API_KEY)                     \
+  do {                                                                     \
+    static auto const is_restricted_api = restricted_apis_.count(API_KEY); \
+    static auto const restricted_api = restricted_apis_.find(API_KEY);     \
```

Comment:
static initialization is now thread safe and happens only once:

https://stackoverflow.com/a/8102145 

the restricted_apis_ map is initialized at start up and then is constant - so no race condition with count and find being out of sync.

Since it's static auto const - it will be initialized once and if is_restricted_api() is false it will be a quick exit (no lookup). 

My hope would be the execution would be smart enough to optimize out the check afterwards via branch prediction ...

Reply:
What I am thinking about is once the program is running for a bit and we have two requests on two different APIs say, inference and model config, and both are checking to see if they are restricted. Is there a race condition in the storage of the `is_restricted_api` variable? Or is this per thread storage?

Reply:
the static const variable is local to the function it is defined - so each function has its own copy that tracks jus that api...

Reply:
💡 Got it! Don't do PRs late at night is what I am learning. The macro defined the `static auto const` in each of the functions thus making the variable easily predicted since it's checking a const string. Sorry for throwing you through the ringer on this. This is better than I first thought.

Reply:
NP, I am wondering now though if the ultimate would be a const local variable initialized from a fixed offset into a table .... I didn't realize the penalty of accessing a static ... may move to a fixed table because we don't need the flexibility of the map (we have a fixed set of categories to restrict)

https://stackoverflow.com/a/21033042